### ISSUE TITLE
AccessViolationException when calling Process.HandleCount

### ISSUE THREAD

**Issue by benaadams** [dotnet/runtime#107503]
### Description

Calling `Process.HandleCount` on a cached handle can cause an `AccessViolationException` on Linux (WSL)

### Reproduction Steps

https://github.com/prometheus-net/prometheus-net/blob/60e9106a83ff1274fec0022c37366f04822b1d1b/Prometheus/DotNetStats.cs#L74-L97

```csharp
public sealed class DotNetStats
{
    private readonly Process _process;
    private Gauge _openHandles;

    private DotNetStats(IMetricFactory metricFactory)
    {
        _process = Process.GetCurrentProcess();
        _openHandles = metricFactory.CreateGauge("process_open_handles", "Number of open handles");
    }
	   
    private readonly object _updateLock = new object();
    public void UpdateMetrics()
    {
        try
        {
            lock (_updateLock)
            {
                _process.Refresh();
				
                _openHandles.Set(_process.HandleCount);
            }
        }
        catch (Exception)
        {
        }
    }
}
```

### Expected behavior

Calling `UpdateMetrics` sets the number of process handles

### Actual behavior

Infrequent (about 6 hours of being called every 5 seconds)
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEntry.Initialize(System.IO.Enumeration.FileSystemEntry ByRef, DirectoryEntry, System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.Span`1<Char>)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(System.Collections.Generic.IEnumerable`1<System.__Canon>)
   at System.Diagnostics.Process.EnsureHandleCountPopulated()
   at Prometheus.DotNetStats.UpdateMetrics()
```

### Regression?

Not sure as first time have run on WSL; haven't had any issues on Windows or seen any complaints from Linux, macOS users

### Known Workarounds

No workaround as `AccessViolationException` always bubbles through catches to crash process

### Configuration

WSL on Windows 11 23H2 (OS Build 22631.4112)

```
.NET SDK:
 Version:           8.0.108
 Commit:            665a05cea7
 Workload version:  8.0.100-manifests.109ff937

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         ubuntu.22.04-x64
 Base Path:   /usr/lib/dotnet/sdk/8.0.108/

.NET workloads installed:
 Workload version: 8.0.100-manifests.109ff937
There are no installed workloads to display.

Host:
  Version:      8.0.8
  Architecture: x64
  Commit:       08338fcaa5

.NET SDKs installed:
  8.0.108 [/usr/lib/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.8 [/usr/lib/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.8 [/usr/lib/dotnet/shared/Microsoft.NETCore.App]
```

### Other information

Aside; creating all the strings for the filenames in `/proc/{pid}/fd` just to find the count of files doesn't seem very efficient

https://github.com/dotnet/runtime/blob/c4792a228ea36792b90f87ddf7fce2477e827822/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs#L172

**Comment by benaadams** [comment]
Though I suppose its more `Directory.GetFiles` on Linux (WSL) than necessarily Process; as that just determines the property of the directory being queried

**Comment by Martin-Molinero** [comment]
Getting a similar errors in linux net9, some stack traces:
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Linq.Enumerable+IEnumerableWhereSelectIterator`2[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Collections.Generic.KeyValuePair`2[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ToArray()

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at Interop+Sys+DirectoryEntry.GetName(System.Span`1<Char>)
   at System.IO.Enumeration.FileSystemEnumerableFactory+<>c__DisplayClass2_0.<UserFiles>b__1(System.IO.Enumeration.FileSystemEntry ByRef)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(System.Collections.Generic.IEnumerable`1<System.__Canon>)
   at System.IO.Directory.GetFiles(System.String, System.String, System.IO.EnumerationOptions)

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Linq.Enumerable+IEnumerableWhereIterator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ToArray()

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Linq.Enumerable+IEnumerableWhereIterator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ToArray()

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Linq.Enumerable+IEnumerableWhereIterator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ToArray()

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(System.Collections.Generic.IEnumerable`1<System.__Canon>)
```

**Comment by jkotas** [comment]
> Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
>   at System.SpanHelpers.NonPackedIndexOfValueType

This looks like some subtle bug in the unsafe code in `NonPackedIndexOfValueType`. cc @MihaZupan @EgorBo 

@Martin-Molinero Any chance you can share a crash dump for this failure? The best way to share crash dump with us is to open an issue on https://developercommunity.visualstudio.com/ and attach the crash dump to it as a private attachment. Unlike github issues, https://developercommunity.visualstudio.com/ does not limits on the size of attachments and the attachments are not publicly visible.

**Comment by MihaZupan** [comment]
Given how commonly used the helper is (`NonPackedIndexOfValueType` is the implementation for `IndexOf(byte)`, I find it suspicious that it's only showing up from `System.IO.Enumeration`.

The only use of `IndexOf(byte)` I see around those stacks is in the Unix Interop `DirectoryEntry` implementation here:
https://github.com/dotnet/runtime/blob/ff604c23939bc59f856ef67f1f3d18dd685ff7a2/src/libraries/Common/src/Interop/Unix/System.Native/Interop.ReadDir.cs#L29-L32
https://github.com/dotnet/runtime/blob/ff604c23939bc59f856ef67f1f3d18dd685ff7a2/src/libraries/Common/src/Interop/Unix/System.Native/Interop.ReadDir.cs#L42-L45

I bet this should be using `MemoryMarshal.CreateReadOnlySpanFromNullTerminated` instead, and we're AVing because the `Name` isn't a full 256 byte buffer.

**Comment by Martin-Molinero** [comment]
Hey! Sorry tried quite a few times to reproduce today enabling the dump generation and guess what, can't reproduce, if I see it again I'll share a dump asap 💯

**Comment by jkotas** [comment]
> we're AVing because the Name isn't a full 256 byte buffer.

How would that happen? Linux has `HAVE_READDIR_R` defined and we should be allocating the buffer of sufficient size ourselves for it here: https://github.com/dotnet/runtime/blob/c9af66ccb24faea7538b682e4d928557276d7367/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs#L48-L49

**Comment by benaadams** [comment]
Reading unaligned vector beyond page size?

Had this issue previously when using IndexOf for searching for null terminators of strings

**Comment by MihaZupan** [comment]
Yep, that's what my guess was, which is resolvable by using `CreateReadOnlySpanFromNullTerminated` instead which ensures that vector reads don't cross page boundaries.
But the current logic that uses `IndexOf` would only be problematic if the span length is incorrect (we'll never read past the length).
To Jan's question, I'm not sure how that could happen yet, but should presumably be visible in a dump.

**Comment by benaadams** [comment]
Had another occurrence https://github.com/NethermindEth/nethermind/issues/8378

Same stack
```
2025-03-15 18:22:06.641	at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware+<InvokeCore>d__4.MoveNext()
2025-03-15 18:22:06.641	at Prometheus.MetricServerMiddleware+<Invoke>d__7.MoveNext()
2025-03-15 18:22:06.641	at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Prometheus.CollectorRegistry+<CollectAndSerializeAsync>d__21, Prometheus.NetStandard, Version=8.0.0.0, Culture=neutral, PublicKeyToken=a243e9817ba9d559]](<CollectAndSerializeAsync>d__21 ByRef)
2025-03-15 18:22:06.641	at Prometheus.CollectorRegistry+<CollectAndSerializeAsync>d__21.MoveNext()
2025-03-15 18:22:06.641	at Prometheus.CollectorRegistry+<RunBeforeCollectCallbacksAsync>d__22.MoveNext()	
2025-03-15 18:22:06.641	at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Prometheus.CollectorRegistry+<RunBeforeCollectCallbacksAsync>d__22, Prometheus.NetStandard, Version=8.0.0.0, Culture=neutral, PublicKeyToken=a243e9817ba9d559]](<RunBeforeCollectCallbacksAsync>d__22 ByRef)
2025-03-15 18:22:06.641	at Prometheus.DotNetStats.UpdateMetrics()
2025-03-15 18:22:06.641	at System.Diagnostics.Process.EnsureHandleCountPopulated()
2025-03-15 18:22:06.641	at System.Collections.Generic.List`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(System.Collections.Generic.IEnumerable`1<System.__Canon>)
2025-03-15 18:22:06.641	at System.IO.Enumeration.FileSystemEnumerator`1[[System.__Canon, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
2025-03-15 18:22:06.641	at System.IO.Enumeration.FileSystemEntry.Initialize(System.IO.Enumeration.FileSystemEntry ByRef, DirectoryEntry, System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.Span`1<Char>)
2025-03-15 18:22:06.641	at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
2025-03-15 18:22:06.641	Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

```

**Comment by benaadams** [comment]
@jkotas @MihaZupan is under `System.Diagnostics.Process.EnsureHandleCountPopulated` which is 

```csharp
string path = Interop.procfs.GetFileDescriptorDirectoryPathForProcess(procPid);
if (Directory.Exists(path))
```

If that makes any difference, might not be preparing the file path in same way?

**Comment by benaadams** [comment]
Which is
```csharp
 internal static string GetFileDescriptorDirectoryPathForProcess(ProcPid pid) =>
    pid == ProcPid.Self ? $"{RootPath}{Self}{FileDescriptorDirectoryName}" :
        string.Create(null, stackalloc char[256], $"{RootPath}{(uint)pid}{FileDescriptorDirectoryName}");
```

**Comment by jkotas** [comment]
I do not think we will be able to make progress on this without a crash dump.

**Comment by Vercidium** [comment]
Hi all, I can reproduce this error reliably. It occurs about 1 in 100 times when starting up my application, so I created a script to repeatedly start+kill my application until it occurs.

The error is always thrown from the Silk.NET `GetUniformLocation` function, which I call hundreds of times on startup:

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.SpanHelpers.NonPackedIndexOfValueType[[System.Byte, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.SpanHelpers+DontNegate`1[[System.Byte, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Byte ByRef, Byte, Int32)
   at System.MemoryExtensions.IndexOf[[System.Byte, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Span`1<Byte>, Byte)
   at Silk.NET.Core.Native.SilkMarshal.Utf8PtrToString(IntPtr)
   at Silk.NET.Core.Native.SilkMarshal.PtrToString(IntPtr, Silk.NET.Core.Native.NativeStringEncoding)
   at Silk.NET.OpenGL.GL.GetUniformLocation(UInt32, System.String)
   at veengine.ShaderProgramGL.GetUniformLocation(System.String)
   ...
```

I'm on Windows 10, .NET 8.0.8.

`SilkMarshal.StringToPtr` allocates the memory using `NativeMemory.Alloc()`:

![Image](https://github.com/user-attachments/assets/e24f39a5-8c3e-419e-98db-014710bd68b6)

Then it calls `Utf8PtrToString()`. In the watch window I've included each character, and it is correctly zero-terminated:

![Image](https://github.com/user-attachments/assets/dc5c703e-e047-493a-bdf7-0ec5409b3d5d)

Unfortunately at this level I can't view any locals variables:

![Image](https://github.com/user-attachments/assets/db1c3e94-58ea-4f7d-85b6-418840b9eea9)

I'll create a ticket on https://developercommunity.visualstudio.com/ now with the crash dump.

**Comment by jkotas** [comment]
@Vercidium `new Span(ptr, Int32.MaxValue).IndexOf<byte>(0)` is not valid way to find the length of null terminated string. The vectorized implementation of IndexOf is free to read memory beyond the terminating null byte. This can lead to intermittent AccessViolationException like the one you are seeing.

The correct way to find the length of null terminated string is using `CreateReadOnlySpanFromNullTerminated` API. I see that the current version of Silk.NET library uses this API: https://github.com/dotnet/Silk.NET/blob/c0dbe835702d01a9f6fe2b5b2ca34f5780d3450d/src/Core/Silk.NET.Core/Native/SilkMarshal.cs#L559 . Perhaps you just need to update to the current version of Silk.NET library to fix this crash in your application.

**Comment by Vercidium** [comment]
Thank you for the quick reply, I can see they fixed this in [v2.21.0](https://github.com/dotnet/Silk.NET/releases/tag/v2.21.0). I'm still on v2.20.0.

I'll update now. Thanks again!

**Comment by jkotas** [comment]
After #116639, we do not use `readdir_r` nor call `IndexOf<byte>(0)` in file enumeration. File enumeration cannot crash with these exact symptoms anymore. 

In case we still see intermittent crashes during file enumeration, let's open a new issue on it.

**Comment by benaadams** [comment]
> After [#116639](https://github.com/dotnet/runtime/pull/116639), we do not use `readdir_r` nor call `IndexOf<byte>(0)` in file enumeration. File enumeration cannot crash with these exact symptoms anymore.

.NET 10?

**Comment by jkotas** [comment]
I do not expect that we will do a full backport of https://github.com/dotnet/runtime/pull/116639. We may backport a reduced version - @jozkee can comment on that.

**Comment by jozkee** [comment]
If we want to be 100% sure about stability, https://github.com/dotnet/runtime/pull/116619 is the best fix to backport, if we ever need to.  Unless you were running in an uncommon filesystem in Unix, like NTFS of ZFS, looks like this issue was mostly caused by the `IndexOf<byte>(0)` on the `byte*`.


I'm wondering if it's possible and worthwhile to have an analyzer that detects the use of IndexOf (and other vectorized operatios) on unmanaged pointers to avoid AVs.

**Comment by jkotas** [comment]
> looks like this issue was mostly caused by the IndexOf<byte>(0) on the byte*.

It was never fully explained what led to this crash - see https://github.com/dotnet/runtime/issues/107503#issuecomment-2637559909 .

**Comment by jozkee** [comment]
Not all Linux configs have HAVE_READDIR_R defined, it is not defined when we build with glibc.

**Comment by MihaZupan** [comment]
> I'm wondering if it's possible and worthwhile to have an analyzer that detects the use of IndexOf (and other vectorized operatios) on unmanaged pointers to avoid AVs.

On its own that could have high false positive rates. There's nothing inherently wrong with using such vectorized helpers on unmanaged memory, issues only appear if you lie about the span's length.
So you could for example flag `new Span(ptr, int.MaxValue)` as potentially problematic, but that wouldn't have helped in this case.

**Comment by jkotas** [comment]
> Not all Linux configs have HAVE_READDIR_R defined, it is not defined when we build with glibc.

Ah ok, I though that it is defined with older glibc that we are building with. It explains it then.
